### ISSUE TITLE
Test ADC response conversion from VDJBase

### ISSUE THREAD

**Issue by bcorrie** [airr-knowledge/issues#67]
Test akc_convert code on AIRR output from VDJBase

- [ ] Support ontology ID (e.g. NCBITAXON:9606) for species.

**Comment by bcorrie** [comment]
@schristley @williamdlees created this issue to track/discuss converting VDJBase AIRR JSON output to LinkML using akc-convert

FYI - AKC convert code is here: https://github.com/airr-knowledge/ak-schema/tree/main/src/scripts/akc-convert

**Comment by bcorrie** [comment]
@williamdlees the following query gives me a response but it isn't "quite" AIRR compliant. 8-)

An AIRR repertoire response should look like this: 
```
{
"Repertoire": [
  {
     "repertoire_id":"repertoire_1", ...
  },
  {
     "repertoire_id":"repertoire_2", ...
  }
]
}
```
The response from your API looks like this:
```
{
  "repertoire": {
    "repertoire_id": "P4_I10_S1", ...
  }
}
```
Differences are the "repertoire" key should be "Repertoire" and it should return an Array of Repertoires, in your case an array of length 1.

Any chance this could be changed 8-)

akc_convert expects all Repertoires to be in the AIRR compliant form.

**Comment by bcorrie** [comment]
The `species` field is required and non-nullable as @schristley says above, but once that is fixed, it just worked (kind of 8-)

Conversion of this query with fixing the AIRR Repertoire stuff above and adding species = Homo sapiens:

https://vdjbase.org/api/v1/airrseq/sample_metadata/Homo%20sapiens/TRB/P4_I10_S1

```
{
  "study_id": {
    "Investigation": {
      "Investigation_BioProject: PRJEB28370": {
        "akc_id": "f3813a07-c19f-4a0b-93f8-6b725cd5bb98",
        "archival_id": "BioProject: PRJEB28370",
        "release_date": "2024-11-22T17:02:20.628677",
        "update_date": "2024-11-22T17:02:20.628686",
        "participants": [
          "c7d7fdde-746a-4135-9b80-d4029d474ec5"
        ],
        "assays": [
          "5bac6807-05b1-4508-ae2e-e1d6684bed33"
        ],
        "documents": [
          "f4b534bc-5a11-4d7e-97bb-da236314e764"
        ],
        "adc_link_tag": "Investigation_BioProject: PRJEB28370",
        "adc_repertoire_id": "P4_I10_S1",
        "adc_study_id": "BioProject: PRJEB28370"
      }
    },
    "StudyArm": {},
    "Reference": {
      "Reference_BioProject: PRJEB28370": {
        "source_uri": "",
        "sources": "https://dx.doi.org/10.3389/fimmu.2018.03004",
        "investigations": [
          "f3813a07-c19f-4a0b-93f8-6b725cd5bb98"
        ],
        "akc_id": "f4b534bc-5a11-4d7e-97bb-da236314e764",
        "adc_link_tag": "Reference_BioProject: PRJEB28370",
        "adc_repertoire_id": "P4_I10_S1",
        "adc_study_id": "BioProject: PRJEB28370"
      }
    },
    "Participant": {
      "Participant_CI10": {
        "akc_id": "c7d7fdde-746a-4135-9b80-d4029d474ec5",
        "species": {
          "id": "NCBITAXON:9606",
          "label": "Homo sapiens"
        },
        "biological_sex": "female",
        "age_event": "f881e1de-f1a6-4b2e-bbf2-ac1ef9e509e1",
        "adc_link_tag": "Participant_CI10",
        "adc_repertoire_id": "P4_I10_S1",
        "adc_study_id": "BioProject: PRJEB28370",
        "adc_subject_id": "CI10",
        "adc_study_group_description": "HCV"
      }
    },
    "ImmuneExposure": {
      "ImmuneExposure_CI10": {
        "akc_id": "1eee0e87-311a-413b-b062-dd6ccf897e7c",
        "life_event": "15bd64f6-000f-435f-a7ac-8aa42ba6f91c",
        "adc_link_tag": "ImmuneExposure_CI10",
        "adc_repertoire_id": "P4_I10_S1",
        "adc_study_id": "BioProject: PRJEB28370",
        "adc_subject_id": "CI10"
      }
    },
    "Specimen": {
      "Specimen_CI10": {
        "akc_id": "21c6d615-ad8a-4dc8-a4b9-6dbc43d6f66e",
        "life_event": "6be7146c-b78f-4575-b92c-4ce1f950bacb",
        "tissue": {
          "label": "blood"
        },
        "adc_link_tag": "Specimen_CI10",
        "adc_repertoire_id": "P4_I10_S1",
        "adc_study_id": "BioProject: PRJEB28370",
        "adc_subject_id": "CI10"
      }
    },
    "LifeEvent": {
      "LifeEvent_Participant_CI10": {
        "akc_id": "f881e1de-f1a6-4b2e-bbf2-ac1ef9e509e1",
        "participant": "c7d7fdde-746a-4135-9b80-d4029d474ec5",
        "adc_link_tag": "LifeEvent_Participant_CI10",
        "adc_repertoire_id": "P4_I10_S1",
        "adc_study_id": "BioProject: PRJEB28370",
        "adc_subject_id": "CI10"
      },
      "LifeEvent_ImmuneExposure_CI10": {
        "akc_id": "15bd64f6-000f-435f-a7ac-8aa42ba6f91c",
        "participant": "c7d7fdde-746a-4135-9b80-d4029d474ec5",
        "adc_link_tag": "LifeEvent_ImmuneExposure_CI10",
        "adc_repertoire_id": "P4_I10_S1",
        "adc_study_id": "BioProject: PRJEB28370",
        "adc_subject_id": "CI10"
      },
      "LifeEvent_Specimen_CI10": {
        "akc_id": "6be7146c-b78f-4575-b92c-4ce1f950bacb",
        "participant": "c7d7fdde-746a-4135-9b80-d4029d474ec5",
        "start": 0,
        "adc_link_tag": "LifeEvent_Specimen_CI10",
        "adc_repertoire_id": "P4_I10_S1",
        "adc_study_id": "BioProject: PRJEB28370",
        "adc_subject_id": "CI10"
      }
    },
    "CellIsolationProcessing": {
      "CellIsolationProcessing_CI10": {
        "akc_id": "277bd086-8de2-4ca2-9c18-db1883c69bb8",
        "specimen": "21c6d615-ad8a-4dc8-a4b9-6dbc43d6f66e",
        "cell_phenotype": "Mixed",
        "single_cell": false,
        "cell_number": 0,
        "cells_per_reaction": 0,
        "cell_storage": false,
        "adc_link_tag": "CellIsolationProcessing_CI10",
        "adc_repertoire_id": "P4_I10_S1",
        "adc_study_id": "BioProject: PRJEB28370",
        "adc_subject_id": "CI10"
      }
    },
    "NucleicAcidProcessing": {
      "NucleicAcidProcessing_CI10": {
        "akc_id": "193543d5-45f2-4a23-9b40-28645df44b92",
        "specimen": "21c6d615-ad8a-4dc8-a4b9-6dbc43d6f66e",
        "template_class": "RNA",
        "template_amount": 0,
        "adc_link_tag": "NucleicAcidProcessing_CI10",
        "adc_repertoire_id": "P4_I10_S1",
        "adc_study_id": "BioProject: PRJEB28370",
        "adc_subject_id": "CI10"
      }
    },
    "LibraryPreparationProcessing": {
      "LibraryPreparationProcessing_CI10": {
        "akc_id": "141991d3-3843-4f99-a5ce-24c66ff5447c",
        "specimen": "21c6d615-ad8a-4dc8-a4b9-6dbc43d6f66e",
        "library_generation_method": "other",
        "adc_link_tag": "LibraryPreparationProcessing_CI10",
        "adc_repertoire_id": "P4_I10_S1",
        "adc_study_id": "BioProject: PRJEB28370",
        "adc_subject_id": "CI10"
      }
    },
    "ReceptorRepertoireSequencingAssay": {
      "ReceptorRepertoireSequencingAssay_CI10": {
        "akc_id": "5bac6807-05b1-4508-ae2e-e1d6684bed33",
        "specimen": "21c6d615-ad8a-4dc8-a4b9-6dbc43d6f66e",
        "type": "ReceptorRepertoireSequencingAssay",
        "complete_sequences": "partial",
        "physical_linkage": "none",
        "adc_link_tag": "ReceptorRepertoireSequencingAssay_CI10",
        "adc_repertoire_id": "P4_I10_S1",
        "adc_study_id": "BioProject: PRJEB28370",
        "adc_subject_id": "CI10"
      }
    }
  },
  "@type": "dict"
}
```

**Comment by bcorrie** [comment]
@williamdlees it looks like the AIRR `sample_id` field is missing. Generating an AKC object from the ADC requires this field. Is this intended? I assumed that this would be the same as the VDJBase sample id, in this case `P4_I10_S1`.

From the AIRR Standard:
```
            type: string
            description: >
                Sample ID assigned by submitter, unique within study. If possible, a persistent sample ID linked to
                INSDC or similar repository study should be used.
            title: Biological sample ID
```

**Comment by bcorrie** [comment]
@williamdlees one other question... If I ask for the genotype for the subject above (`CI10`) with this query:
```
curl -X 'GET' \
  'https://vdjbase.org/api/v1/airrseq/subject_genotype/Homo%20sapiens/CI10' \
  -H 'accept: application/json'
```
I am getting this result: 
```
Error: response status is 400
Response body: {"message":"Subject not found"}
```

**Comment by schristley** [comment]
Thanks @bcorrie for testing this out. Note that we won't be converting VDJbase repertoires to AKC investigations and such. All of these repertoires should essentially already be in the ADC (given some time lag). What we will want to do is bring in the `Genotype` from VDJbase and link to an existing AKC Participant and Investigation by using the appropriate identifiers.

**Comment by bcorrie** [comment]
> All of these repertoires should essentially already be in the ADC (given some time lag).

This is assuming that VDJBase would never have information about studies that are not in the ADC. Is that correct?

For example, wouldn't studies that have genomic genotypes be of interest to the AKC, and these may not have AIRR-seq data in the ADC.

**Comment by bcorrie** [comment]
> What we will want to do is bring in the `Genotype` from VDJbase and link to an existing AKC Participant and Investigation by using the appropriate identifiers.

This relates to our discussion yesterday (and #56) in terms of linking back and forth between the AKC and the component repositories - how are we going to do that?

Currently in the AKC, we do not retain any of the ADC identifiers in the AKC. There is no equivalent of `subject_id` and `sample_id`.

So if I get the Genotype for a `Subject` in VDJBase, and that `Participant` has already been created in the AKC from `Subject` data in the ADC, how will I determine which `Participant` that is in the AKC? Currently, with the fields we have, this is not possible.

Conversely, if I am in the AKC world, how would I figure out what the source data for the AKC `Participant` is. A given `Participant` in the AKC might have subject metadata from the ADC and subject Genotype from the VDJBase. The AKC needs to track this for reproducibility and data provenance.

**Comment by williamdlees** [comment]
OK, that's easy to fix.


------ Original Message ------
From "Brian Corrie" ***@***.***>
To "airr-knowledge/issues" ***@***.***>
Cc "William Lees" ***@***.***>; "Mention" 
***@***.***>
Date 22/11/2024 17:30:19
Subject Re: [airr-knowledge/issues] Test ADC response conversion from 
VDJBase (Issue #67)

**Comment by williamdlees** [comment]
In line with the definiton you quote, we take this to be the sample ID 
issued by the original researcher. It will be there wherever it's known. 
It should be there for all the projects for which we have downloaded 
information from the ADC. Information for the others is pretty sparse, 
as you have probably found. There is work going on to correct that. If 
we establish that there really weren't any IDs issued, for example for 
one of Gur's own projects, we can copy the VDJbase sample_id over.


------ Original Message ------
From "Brian Corrie" ***@***.***>
To "airr-knowledge/issues" ***@***.***>
Cc "William Lees" ***@***.***>; "Mention" 
***@***.***>
Date 22/11/2024 18:29:00
Subject Re: [airr-knowledge/issues] Test ADC response conversion from 
VDJBase (Issue #67)

**Comment by williamdlees** [comment]
It's expecting a VDJbase subject, e.g.

curl -X 'GET' \
   'https://vdjbase.org/api/v1/airrseq/subject_genotype/Homo%20sapiens/P4_I1' \
   -H 'accept: application/json'


------ Original Message ------
From "Brian Corrie" ***@***.***>
To "airr-knowledge/issues" ***@***.***>
Cc "William Lees" ***@***.***>; "Mention" 
***@***.***>
Date 22/11/2024 18:33:33
Subject Re: [airr-knowledge/issues] Test ADC response conversion from 
VDJBase (Issue #67)

**Comment by bcorrie** [comment]
> It's expecting a VDJbase subject, e.g.

Can I get the VDJBase subject from the VDJBase AIRR response? It doesn't appear that this information is returned.

This would mean that I can't do a query against VDJBase for subject genotype using data that is in the AKC that was generated from data in the ADC.

**Comment by williamdlees** [comment]
I guess not, because the field isn't in the AIRR standard! And sample_id 
is optional...

I stopped doing too much AIRR-centric development of the API because 
Scott felt that the IAKC would probably take data in bulk to keep the 
load on both systems reduced. You can ask for all the metadata with 
all_subjects_metadata and all the genotypes with all_subjects_genotypes. 
It takes a while the first time but the response is cached until the 
data changes, so it's relatively low impact.

Happy to add something more granular in if it's important, but this was 
the thinking.

------ Original Message ------
From "Brian Corrie" ***@***.***>
To "airr-knowledge/issues" ***@***.***>
Cc "William Lees" ***@***.***>; "Mention" 
***@***.***>
Date 22/11/2024 19:37:57
Subject Re: [airr-knowledge/issues] Test ADC response conversion from 
VDJBase (Issue #67)

**Comment by schristley** [comment]
> > All of these repertoires should essentially already be in the ADC (given some time lag).
> 
> This is assuming that VDJBase would never have information about studies that are not in the ADC. Is that correct?
> 
> For example, wouldn't studies that have genomic genotypes be of interest to the AKC, and these may not have AIRR-seq data in the ADC.

Correct. Maybe with genomic genotypes but if we don't have chain/receptor information as measured through an AIRR-seq experiment, then the genomic genotype isn't very useful to the AKC.

**Comment by williamdlees** [comment]
Even if we have disease/response information, for example?


------ Original Message ------
From "Scott Christley" ***@***.***>
To "airr-knowledge/issues" ***@***.***>
Cc "William Lees" ***@***.***>; "Mention" 
***@***.***>
Date 22/11/2024 19:48:41
Subject Re: [airr-knowledge/issues] Test ADC response conversion from 
VDJBase (Issue #67)

**Comment by schristley** [comment]
> > What we will want to do is bring in the `Genotype` from VDJbase and link to an existing AKC Participant and Investigation by using the appropriate identifiers.
> 
> This relates to our discussion yesterday (and #56) in terms of linking back and forth between the AKC and the component repositories - how are we going to do that?
> 
> Currently in the AKC, we do not retain any of the ADC identifiers in the AKC. There is no equivalent of `subject_id` and `sample_id`.

You might have missed my #55 response [here](https://github.com/airr-knowledge/issues/issues/55#issuecomment-2452901516). Those user-supplied IDs can be put in the AKC `name` field.

**Comment by bcorrie** [comment]
I am getting:
```
curl -X 'GET'    'https://vdjbase.org/api/v1/airrseq/all_subjects_genotype/Homo%20sapiens'    -H 'accept: application/json'
<!doctype html>
<html lang=en>
<title>500 Internal Server Error</title>
<h1>Internal Server Error</h1>
<p>The server encountered an internal error and was unable to complete your request. Either the server is overloaded or there is an error in the application.</p>
```

**Comment by williamdlees** [comment]
Sorry, I can see there's an issue with the 'get all' api - will look at 
it this weekend


------ Original Message ------
From "William Lees" ***@***.***>
To "airr-knowledge/issues" 
***@***.***>
Date 22/11/2024 19:45:36
Subject Re[2]: [airr-knowledge/issues] Test ADC response conversion from 
VDJBase (Issue #67)

**Comment by williamdlees** [comment]
Yes, it sent me the stack trace. There's a bug in the cacheing code.


------ Original Message ------
From "Brian Corrie" ***@***.***>
To "airr-knowledge/issues" ***@***.***>
Cc "William Lees" ***@***.***>; "Mention" 
***@***.***>
Date 22/11/2024 19:54:29
Subject Re: [airr-knowledge/issues] Test ADC response conversion from 
VDJBase (Issue #67)

**Comment by schristley** [comment]
> Even if we have disease/response information, for example?

I won't completely discount it, but we have multiple things to resolve before we can incorporate genomic genotypes. We don't have a standardized description of the experimental protocol, as the AIRR standards is specifically for AIRR-seq. We haven't decided how to do linking of data in the AKC which isn't based on a chain/receptor. And there is probably more information and/or different information in the genomic experiment (I assume this is why they have completely separate endpoints, otherwise just a flag would be sufficient). So I think it is worthwhile to consider and resolve those issues, but not for this interim data integration prototype.

**Comment by bcorrie** [comment]
> You might have missed my #55 response [here](https://github.com/airr-knowledge/issues/issues/55#issuecomment-2452901516). Those user-supplied IDs can be put in the AKC `name` field.

I did miss that, thanks.

None of that is in the current data model. In fact, in the AKC LinkML schema there is nowhere to capture that such a mapping is desired. The only place where we can capture that mapping is in the Consolidated Mapping google sheet - which should remind us how important it is and that we have to keep it up to date... 8-)

**Comment by bcorrie** [comment]
> OK, that's easy to fix.

@williamdlees I think the all_samples_metadata API endpoint (at least according to the OpenAPI web interface) will generate a list of repertoires with the following format:
```
{
  "repertoire_class_list": [
    "string"
  ]
}
```
If that list of strings is a list of repertoires, then this should be:
```
{
  "Repertoire": [
    {"repertoire_id":"repertoire_1", ... },{"repertoire_id":"repertoire_2", ... }, ...
  ]
}
```

**Comment by williamdlees** [comment]
Brian,

The two 'bulk' queries are working ok now, I think.

Regarding the format of the bulk metadata download, the top-level 
structure at the moment is:

{
   repertoire_class_list: [
     ....
  ]
}

I think everything underlying is as it should be, but 
'repertoire_class_list' should be replaced with the text 'Repertoire'

Perhaps you can have a look now the call is working and see if you 
agree.

Thanks

William

------ Original Message ------
From "Brian Corrie" ***@***.***>
To "airr-knowledge/issues" ***@***.***>
Cc "William Lees" ***@***.***>; "Mention" 
***@***.***>
Date 22/11/2024 19:54:29
Subject Re: [airr-knowledge/issues] Test ADC response conversion from 
VDJBase (Issue #67)

**Comment by bcorrie** [comment]
> I think everything underlying is as it should be, but 'repertoire_class_list' should be replaced with the text 'Repertoire'

Yes, the AIRR Standard states that a list of `Repertoire` objects should be of the form:
```
{"Repertoire":[]}
```

**Comment by bcorrie** [comment]
@williamdlees the best way to test compliance of VDJBase AIRR output is to probably just dump the data to a AIRR JSON file and load it into python using the AIRR python library as this will complain if anything is not according to the AIRR Standard.

The AIRR to AKC conversion code is here: 

https://github.com/airr-knowledge/ak-schema/tree/specimen-processing/src/scripts/akc-convert

You have to run it within the AKC docker container.

If you give it a list of Repertoires as spec'ed above, it should convert the output to AKC compliant linkML.

This would be a good test if the VDJBase AIRR output is standard compliant, but if something is not AIRR compliant akc_convert will fail with errors and not tell you what is wrong with the AIRR standard output. The AIRR python library will provide more useful information from a standards compliance perspective.

**Comment by williamdlees** [comment]
I finally found some time to change the attribute to Repertoire: and the json loads cleanly as below.

```
digby_genomics310) d:\Research\ikc>curl https://vdjbase.org/api/v1/airrseq/sample_metadata/Homo%20sapiens/IGH/P1_I1_S1 -o P1_I1_S1_metadata.json
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100  3201  100  3201    0     0   6690      0 --:--:-- --:--:-- --:--:--  6738

(digby_genomics310) d:\Research\ikc>curl https://vdjbase.org/api/v1/airrseq/subject_genotype/Homo%20sapiens/P1_I1 -o P1_I1_genotype.json
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100 23296  100 23296    0     0  28996      0 --:--:-- --:--:-- --:--:-- 29120

(digby_genomics310) d:\Research\ikc>curl https://vdjbase.org/api/v1/airrseq/all_samples_metadata/Homo%20sapiens/IGH -o IGH_metadata.json
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100 6732k  100 6732k    0     0   616k      0  0:00:10  0:00:10 --:--:-- 1534k

(digby_genomics310) d:\Research\ikc>curl https://vdjbase.org/api/v1/airrseq/all_subjects_genotype/Homo%20sapiens -o airr_genotypes.json
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100 13.9M  100 13.9M    0     0  3167k      0  0:00:04  0:00:04 --:--:-- 3402k

(digby_genomics310) d:\Research\ikc>ls -l
total 21088
-rw-r--r-- 1 William 197121  6894441 Dec  9 15:30 IGH_metadata.json
-rw-r--r-- 1 William 197121     3201 Dec  9 15:30 P1_I1_S1_metadata.json
-rw-r--r-- 1 William 197121    23296 Dec  9 15:30 P1_I1_genotype.json
-rw-r--r-- 1 William 197121 14666279 Dec  9 15:30 airr_genotypes.json

(digby_genomics310) d:\Research\ikc>python
Python 3.10.8 | packaged by conda-forge | (main, Nov 24 2022, 14:07:00) [MSC v.1916 64 bit (AMD64)] on win32
Type "help", "copyright", "credits" or "license" for more information.
>>> import airr
>>> data = airr.read_airr('P1_I1_S1_metadata.json')
>>> data = airr.read_airr('P1_I1_genotype.json')
>>> data = airr.read_airr('IGH_metadata.json')
>>> data = airr.read_airr('airr_genotypes.json')
>>> quit()
```

**Comment by schristley** [comment]
VDJbase integration script has been written and tested so the repertoire data should be good.
